launcher/ui/themes/ThemeManager.cpp

```diff
@@ -0,0 +1,133 @@
+    return theme->id();
+}
+
+/// @brief Gets the Theme from the List via ID
+/// @param themeId Theme ID of theme to fetch
+/// @return Theme at themeId
+ITheme* ThemeManager::GetTheme(QString themeId) {
+    return m_themes[themeId].get();
+}
+
+void ThemeManager::InitializeThemes() {
+    
+
+    // Icon themes
+    {
+        // TODO: icon themes and instance icons do not mesh well together. Rearrange and fix discrepancies!
+        // set icon theme search path!
+        auto searchPaths = QIcon::themeSearchPaths();
+        searchPaths.append("iconthemes");
+        QIcon::setThemeSearchPaths(searchPaths);
+        themeDebugLog << "<> Icon themes initialized.";
+    }
+
+    // Initialize widget themes
+    {
+        themeDebugLog << "<> Initializing Widget Themes";
+        themeDebugLog "✓ Loading Built-in Theme:" << AddTheme(new SystemTheme());
+        auto darkThemeId = AddTheme(new DarkTheme());
+        themeDebugLog "✓ Loading Built-in Theme:" << darkThemeId;
+        themeDebugLog "✓ Loading Built-in Theme:" << AddTheme(new BrightTheme());
+
+        // TODO: need some way to differentiate same name themes in different subdirectories (maybe smaller grey text next to theme name in dropdown? Dunno how to do that though)
+        QString themeFolder = (new QDir("./themes/"))->absoluteFilePath("");
+        themeDebugLog << "Theme Folder Path: " << themeFolder;
+
+        QDirIterator directoryIterator(themeFolder, QDir::Dirs | QDir::NoDotAndDotDot, QDirIterator::Subdirectories);
+        while (directoryIterator.hasNext()) {
+            QDir dir(directoryIterator.next());
+            QFileInfo themeJson(dir.absoluteFilePath("theme.json"));
+            if (themeJson.exists()) {
+                // Load "theme.json" based themes
+                themeDebugLog "✓ Loading JSON Theme from:" << themeJson.absoluteFilePath();
+                CustomTheme* theme = new CustomTheme(GetTheme(darkThemeId), themeJson, true);
+                AddTheme(theme);
+            } else {
+                // Load pure QSS Themes
+                QDirIterator stylesheetFileIterator(dir.absoluteFilePath(""), {"*.qss", "*.css"}, QDir::Files);
+                while (stylesheetFileIterator.hasNext()) {
+                    QFile customThemeFile(stylesheetFileIterator.next());
+                    QFileInfo customThemeFileInfo(customThemeFile);
+                    themeDebugLog "✓ Loading QSS Theme from:" << customThemeFileInfo.absoluteFilePath();
+                    CustomTheme* theme = new CustomTheme(GetTheme(darkThemeId), customThemeFileInfo, false);
+                    AddTheme(theme);
+                }
+            }
+        }
+
+        themeDebugLog << "<> Widget themes initialized.";
+    }
+}
+
+std::vector<ITheme *> ThemeManager::getValidApplicationThemes()
+{
+    std::vector<ITheme *> ret;
+    auto iter = m_themes.cbegin();
+    while (iter != m_themes.cend())
+    {
+        ret.push_back((*iter).second.get());
+        iter++;
+    }

```

Comment:
this part was code I had moved from Application.cpp, so I haven't actually looked into optimizing it, I will have a look at the two links you sent though and see what I can do ^^

Reply:
I've changed the loop to be a "for each", this seems to compile and run fine, but I don't actually know what the functions in the loop did or do now. 
Like.. I know it puts the Themes into a "List" (Vector), but I don't actually know what e.g. emplace does xD
```cpp
std::vector<ITheme *> ThemeManager::getValidApplicationThemes()
{
    std::vector<ITheme *> ret;
    ret.reserve(m_themes.size());
    for (auto&& [id, theme] : m_themes) {
        ret.emplace_back(theme);
    }
    return ret;
}
```

Reply:
this is called a range-for loop :)
emplace construct the element in place in the vector instead of making a copy. it forwards the argument until it can construct the element in the vector directly.

btw, it looks strange that it compiles as is, as theme should be a `std::unique_ptr`, so you would need to do `.get()` on it to have a raw pointer

Reply:
in this case, emplace is not really doing anything different from push_back, as you are putting raw pointers in the vector

Reply:
I will add the .get(), but it seemed to work fine without it too.